PublicAPI/Microsoft.ApplicationInsights.dll/net45/PublicAPI.Unshipped.txt

```diff
@@ -33,5 +33,237 @@ Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.IHeartbeatPro
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name) -> Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name, Microsoft.ApplicationInsights.Metrics.MetricConfiguration metricConfiguration) -> Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name, Microsoft.ApplicationInsights.Metrics.MetricConfiguration metricConfiguration, Microsoft.ApplicationInsights.MetricAggregationScope aggregationScope) -> Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name, string dimension3Name) -> Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name, string dimension3Name, Microsoft.ApplicationInsights.Metrics.MetricConfiguration metricConfiguration) -> Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name, string dimension3Name, Microsoft.ApplicationInsights.Metrics.MetricConfiguration metricConfiguration, Microsoft.ApplicationInsights.MetricAggregationScope aggregationScope) -> Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name, string dimension3Name, string dimension4Name) -> Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name, string dimension3Name, string dimension4Name, Microsoft.ApplicationInsights.Metrics.MetricConfiguration metricConfiguration) -> Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.TelemetryClient.GetMetric(string metricId, string dimension1Name, string dimension2Name, string dimension3Name, string dimension4Name, Microsoft.ApplicationInsights.Metrics.MetricConfiguration metricConfiguration, Microsoft.ApplicationInsights.MetricAggregationScope aggregationScope) -> Microsoft.ApplicationInsights.Metric
+const Microsoft.ApplicationInsights.MetricDimensionNames.TelemetryContext.Cloud.RoleInstance = "TelemetryContext.Cloud.RoleInstance" -> string
+const Microsoft.ApplicationInsights.MetricDimensionNames.TelemetryContext.Cloud.RoleName = "TelemetryContext.Cloud.RoleName" -> string
+const Microsoft.ApplicationInsights.MetricDimensionNames.TelemetryContext.Component.Version = "TelemetryContext.Component.Version" -> string
+const Microsoft.ApplicationInsights.MetricDimensionNames.TelemetryContext.Device.Id = "TelemetryContext.Device.Id" -> string
```

Comment:
why do you need all these constants exposed publicly? Especially I concerned for the deprecated properties like device ID and screen resolution

Reply:
Thank you for this question.
The usage, the reason and the background for the constants in `MetricDimensionNames` are described in detail in the doc comments for that class, as well as in the samples 2 and 5:

https://github.com/Microsoft/ApplicationInsights-dotnet/blob/feature/PreAggMetrics.Master/Test/Microsoft.ApplicationInsights.Test/Shared/Metrics/MetricsExamples.cs

We have also previously discussed this, albeit briefly.

Please let me re-iterate some key background points to put the aforementioned resources in context:

A key feedback point for MetricManager v1 was that it did not allow to specify `TelemetryContext` properties. For example, if other kinds of telemetry are associated with a specific *session* or a specific *operation name*, they use the corresponding properties within the `TelemetryContext` to convey that information. However, metric aggregates produced by MetricManager v1 did allow access to the `TelemetryContext`, and such information needed to be stored as a normal dimensions. The inability to overcome this inconsistency was named as a release blocker by the SDK team at that time. 

Exposing `TelemetryContext` for metric series remains impractical because series represent aggregation scopes and require immutable identities that are fast to compare. So, to provide a solution for the described problem, we use well-defined dimension names. Dimensions with such names are treated as normal dimensions during pre-aggregation, but when the resulting aggregates are serialized for transport, the values of such dimensions are stored within the corresponding fields of the `TelemetryContext`.

Please let me know if this answers your question.
If you require adjustments to these APIs, please request *precisely* the changes you would like to see so that I can make them. Especially, if certain fields do no longer belong into the `TelemetryContext` and should be treated as normal dimensions, we should remove them from the constants list.
Thank you.